### DISCUSSION TITLE
Best way to .gitignore all code but my own inside a devcontainer?

### DISCUSSION THREAD

**Discussion by C-Loftus** [devcontainers/community#164] [Q&A]
## Question

What is the best way to use git inside a devcontainer if you are developing an addon for a base image where the base image should not be changed?

## Background

- I write plugins for a Python server that are packaged independently of the server image
- I need to be inside the container to get the proper environment for code completion
- I mount all my work into the container so it is separated and persistent, but still accessible via the container
- I mount my `.git/` folder and push my code via the git cli 

Ideally I would love to be able to make the base container read only and have it be ignored by git. 

## Challenge

- When I develop inside the container, git tries to track changes for all the files from the base container image, not just my code
- I can use a `.gitignore` like the following, but then I need to manually specify each file I want to not ignore
    - I feel it is error prone and a bad practice to ignore everything by default

```
*

!.devcontainer/
!.vscode/
# need to use some hacky recursive unignores
!src/
!src/**

!.gitignore
```

- I also need to end up mounting every file I want to track, otherwise, it shows up as deleted in git when I enter the container 
    - (since it wasn't mounted and thus isn't in the container filesystem)
    - this can be an issue if one if my files shares the same name as one in the container (i.e. readme.md or dockerfile)
    - this makes my `devcontainer.json` particularly verbose
    - combined with the verbose `.gitignore`, I end up having to append to two files every time I want to add a new file at the root of the workspace

**Comment by C-Loftus** [comment]
It appears that an overlay filesystem may be appropriate. However, I don't believe that overlay filesystem mounts are supported by devcontainers. Perhaps I am wrong though. Still a bit stuck on this

**Comment by null-hype** [comment]
Stop me if I'm changing the subject.

You've probably seen [`optionalPaths`](https://github.com/microsoft/vscode-docs/blob/main/remote-release-notes/v1_93.md#optional-paths-for-templates).

![optional paths UX](https://media.githubusercontent.com/media/microsoft/vscode-docs/main/remote-release-notes/images/1_93/optionalPaths.png)

One could `devcontainer build` and then include a `devcontainer-template.json` which explicitly defines the `optionalPaths` to be included in the build. This would be useful for scenarios where you want to include certain files in the build, but not others.

You could then `devcontainer templates apply` and use the `--omit-paths` option to exclude certain files from the build. This would allow you to start from the base image, and apply only those changes you are concerned with before commiting. All with nice UI (if you're uing VSCode.)
